File: solutions_1483485_0/java/alterloki/QualificationA.java
/**
 * Created by IntelliJ IDEA.
 * User: ashevenkov
 * Date: 14.04.12
 * Time: 11:17
 */
package codejam.codejam2012;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.*;

/**
 * @author ashevenkov
 */
public class QualificationA {

    private Map<String, String> inputToOutput = new HashMap<String, String>();
    private Map<Character, Character> mapping = new TreeMap<Character, Character>();

    public QualificationA() {
        inputToOutput.put(
                "ejp mysljylc kd kxveddknmc re jsicpdrysi",
                "our language is impossible to understand");
        inputToOutput.put(
                "rbcpc ypc rtcsra dkh wyfrepkym veddknkmkrkcd",
                "there are twenty six factorial possibilities");
        inputToOutput.put(
                "de kr kd eoya kw aej tysr re ujdr lkgc jv",
                "so it is okay if you want to just give up");
        init();
    }

    private void init() {
        for(Map.Entry<String, String> entry : inputToOutput.entrySet()) {
            char[] inCa = entry.getKey().toCharArray();
            char[] outCa = entry.getValue().toCharArray();
            for(int i = 0; i < inCa.length; i++) {
                char in = inCa[i];
                char out = outCa[i];
                if(in != ' ') {
                    mapping.put(in, out);
                }
            }
        }
        mapping.put('q', 'z');
        mapping.put('z', 'q');
    }

    public static void main(String[] args) throws Exception {
        new QualificationA().calculate(
                "/home/ashevenkov/projects/codejam/2012-QualificationA.in",
                "/home/ashevenkov/projects/codejam/2012-QualificationA.out");
    }

    public void calculate(String in, String out) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader(in));
        FileWriter fw = new FileWriter(out);
        String line = br.readLine();
        int cases = Integer.parseInt(line);
        for(int i = 0; i < cases; i++) {
            line = br.readLine();
            String result = translate(line);
            System.out.println("Case #" + (i + 1) + ": " + result);
            fw.write("Case #" + (i + 1) + ": " + result);
            fw.write("\n");
        }
        fw.flush();
        fw.close();
    }

    private String translate(String line) {
        StringBuilder sb = new StringBuilder();
        char[] ca = line.toCharArray();
        for(int i = 0; i < ca.length; i++) {
            char c = ca[i];
            if(c != ' ') {
                sb.append(mapping.get(c));
            } else {
                sb.append(' ');
            }
        }
        return sb.toString();
    }

}

File: solutions_1483485_0/C++/Ansel/QualificationA.cpp
#include <assert.h>
#include <string.h>

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

const char * ciphertext[3] = {
		"ejp mysljylc kd kxveddknmc re jsicpdrysi",
		"rbcpc ypc rtcsra dkh wyfrepkym veddknkmkrkcd",
		"de kr kd eoya kw aej tysr re ujdr lkgc jv"
};

const char * plaintext[3]  = {
		"our language is impossible to understand",
		"there are twenty six factorial possibilities",
		"so it is okay if you want to just give up",
};

// cipherchar : index
// plainchar  : key
char maptable[26];

void init_maptable ()
{
	int	i, j;
	for (i = 0; i < 3; ++i) {
		int length = strlen(ciphertext[i]);
		for (j = 0; j < length; ++j) {
			char cipherchar = ciphertext[i][j];
			char plainchar  = plaintext[i][j];
			if (cipherchar != ' ') {
				maptable[cipherchar - 'a'] = plainchar;
			}
		}
	}
	maptable['z'-'a'] = 'q';
	maptable['q' - 'a'] = 'z';
//	for (i = 0; i < 26; ++i) {
//		cout << "cipherchar = " << char(i + 'a') << "; plainchar = " << maptable[i] << endl;
//	}
//	for (i = 0; i < 25; ++i) {
//		for (j = i+1; j < 26; ++j) {
//			if (maptable[j] == maptable[i]) {
//				cout << "Error:" << "i = " << i << "; j = " << j << "; maptable[] = " << maptable[i] << endl;
//				exit(1);
//			}
//		}
//	}
//	cout << "Right" << endl;
}

void decode(const char* cipherline,  char * plainline)
{
	int length = strlen(cipherline);
	for ( ; *cipherline != '\0'; plainline++, cipherline++) {
		if (*cipherline >= 'A' && *cipherline <= 'Z') {
			*plainline = maptable[*cipherline - 'A'];
		} else if (*cipherline >= 'a' && *cipherline <= 'z') {
			*plainline = maptable[*cipherline - 'a'];
		} else {
			*plainline = *cipherline;
		}
	}
	*plainline = *cipherline;
}

int main() {
	init_maptable();

	int 	N;
	char	line[256];
	char	plainline[256];

	cin.getline(line, 256);
	N = atoi(line);
	assert(N > 0);

	for (int i = 0; i < N; i++) {
		cin.getline(line, 256);
		int length = strlen(line);
		assert(length > 0 && length <= 100);
		decode(line, plainline);
		cout << "Case #" << i+1 << ": " << plainline << endl;
	}

	return 0;
}
